### PR TITLE
Support constraints for `Base64Bytes` with core schema fix

### LINKED ISSUE #9251: Base64Bytes field doesn't raises validation error for min_length constraint
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

The validation should raise an error due to the minimum length constraint set for the `data` field in the `Test` model. However, no validation error is raised even though the `data` field is empty.

### Expected Behavior:
Passing empty data for the `data` field should raise a validation error due to the minimum length constraint set in the `Test` model.

### Actual Behavior:
No validation error is raised when passing empty data for the `data` field.

### Steps to Reproduce:

Define the provided models in a Pydantic-based Python script.
Attempt to validate input data where the `data` field is empty.
Notice that no validation error is raised even though the `data` field does not meet the minimum length constraint.
Additional Information:
This behavior was observed in Pydantic version `2.6.4`.

### Example Code

```Python
from pydantic import Base64Bytes, BaseModel, Field


class Test(BaseModel):
    data: Base64Bytes = Field(min_length=10)


test = Test(data=b"")  # no validation error
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.6.4
pydantic-core version: 2.16.3
pydantic-core build: profile=release pgo=true
python version: 3.10.11 (tags/v3.10.11:7d4cc5a, Apr  5 2023, 00:38:17) [MSC v.1929 64 bit (AMD64)]
platform: Windows-10-10.0.22631-SP0
```

### PR DESCRIPTION
**Pull request by fatelei**
## Change Summary

enhance base64 bytes check.

Fix #9251 

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: pydantic/types.py [hunk 1]
```diff
@@ -2393,9 +2393,12 @@ def __get_pydantic_json_schema__(
         return field_schema
 
     def __get_pydantic_core_schema__(self, source: type[Any], handler: GetCoreSchemaHandler) -> core_schema.CoreSchema:
+        original_schema = handler(source)
+        # bytes schema has no keyword argument type.
+        original_schema.pop("type", None)
         return core_schema.with_info_after_validator_function(
             function=self.decode,
-            schema=core_schema.bytes_schema(),
+            schema=core_schema.bytes_schema(**original_schema),
```
Comment (Viicos) [inline pydantic/types.py:2401]: Can't we just do:

```suggestion
        return core_schema.with_info_after_validator_function(
            function=self.decode,
            schema=handler(source),
        )
```

and maybe additionally raise if `handler(source)` is not a bytes schema?

Comment (fatelei) [reply pydantic/types.py]: update it

File: pydantic/types.py [hunk 2]
```diff
@@ -2393,9 +2393,11 @@ def __get_pydantic_json_schema__(
         return field_schema
 
     def __get_pydantic_core_schema__(self, source: type[Any], handler: GetCoreSchemaHandler) -> core_schema.CoreSchema:
+        if isinstance(source(), bytes):
+            raise TypeError("source is not a bytes schema")
```
Comment (Viicos) [inline pydantic/types.py:2397]:
```suggestion
            raise TypeError(f"{type(self).__name__} should only be applied on bytes, got {source}")
```

Comment (fatelei) [reply pydantic/types.py]: update

File: pydantic/types.py [hunk 3]
```diff
@@ -2393,9 +2393,11 @@ def __get_pydantic_json_schema__(
         return field_schema
 
     def __get_pydantic_core_schema__(self, source: type[Any], handler: GetCoreSchemaHandler) -> core_schema.CoreSchema:
+        if isinstance(source(), bytes):
```
Comment (sydney-runkle) [inline pydantic/types.py:2396]: `if not`?

Comment (sydney-runkle) [reply pydantic/types.py:2396]: I feel like we should do something more like the following, to be more consistent with other types in this file...

```py
if cls is source:
    # used directly as a type
     schema = core_schema.bytes_schema()
else:
    schema = handler(source)
    _check_annotated_type(schema['type'], 'bytes', cls.__name__)

return core_schema.with_info_after_validator_function(
    function=self.decode,
    schema=schema,
    serialization=core_schema.plain_serializer_function_ser_schema(function=self.encode),
)
```

**Review by sydney-runkle — COMMENTED** [review]
Looking good overall - left one comment. Tests are failing as well, so we'll want to get that sorted.

File: pydantic/types.py (hunk 3 above)
Comment (Viicos) [reply pydantic/types.py:2396]: Is the first condition possible? I don't think it makes sense to use `EncodedBytes` as is: `TypeAdapter(EncodedBytes(...))`

Comment (fatelei) [reply pydantic/types.py]: maybe 

```
schema = handler(source)
 _check_annotated_type(schema['type'], 'bytes', self.__classs__.__name__)
```

is enough

**Review by sydney-runkle — APPROVED** [review]
Looks great, thanks for the help with the fix!

File: pydantic/types.py [hunk 4]
```diff
@@ -2476,9 +2476,12 @@ def __get_pydantic_json_schema__(
         return field_schema
 
     def __get_pydantic_core_schema__(self, source: type[Any], handler: GetCoreSchemaHandler) -> core_schema.CoreSchema:
+        schema = handler(source)
+        _check_annotated_type(schema['type'], 'bytes', self.__class__.__name__)
+
```
Comment (sydney-runkle) [inline pydantic/types.py:2481]:
```suggestion

```

**Comment by sydney-runkle** [comment]
Hmm, still experiencing test failures here. @fatelei, could you please take a look at those?

**Comment by fatelei** [comment]
because EncodedStr inhert EncodedBytes, so i remove the type check

```
_check_annotated_type(schema['type'], 'bytes', self.__classs__.__name__)
```

because EncodedStr can't pass this check.

and i fixed the doc test of EncodedStr

**Comment by fatelei** [comment]
ut has passed, but i am not sure how to fix lint error

image

**Review by sydney-runkle — APPROVED** [review]
I think this should suffice, thanks for your help!

**Comment by sydney-runkle** [comment]
Decided on:

```py
schema = handler(source)
        if (schema_type := schema['type']) not in ('bytes', 'str'):
            raise PydanticUserError(
                f"'EncodedBytes' cannot annotate '{schema_type}'.", code='invalid-annotated-type'
            )
```

Because `isinstance` supports subclasses, and `EncodedStr` is a subclass of `EncodedBytes`, which doesn't play well with this. We should probably just not use `super()` for core schema gen in `EncodedStr`. Happy to address that in a follow up PR.

File: pydantic/types.py [hunk 5]
```diff
@@ -2411,6 +2411,8 @@ def get_json_format(cls) -> Literal['base64url']:
         return 'base64url'
 
 
+# TODO: align with PastDate and other similar types to enable usage alone
+# and add an if TYPE_CHECKING block with Annotated[bytes, ...] etc to make type checkers happy
```
Comment (sydney-runkle) [inline pydantic/types.py:2415]: Following up - I don't think this makes sense, as `EncodedBytes` takes an encoder argument, so it doesn't really make sense to initialize independently, I think we can leave the structure as is.

**Comment by sydney-runkle** [comment]
@fatelei,

I'll experiment a bit more with this today, we'll take it from here! Thanks for your help.

**Comment by sydney-runkle** [comment]
Closing this, I've implemented a slightly different fix in #10584.

**Pull request closed without merging**